### PR TITLE
cpu/samd21: Add samd21 pwm driver

### PR DESCRIPTION
**Pull request by PeterKietzmann**
This is an implementation of a pwm lowlevel driver for samr21-xpro boards. Please note that the `pwm_mode_t mode` is not yet implemented for this board.

### CODE REVIEW

File: cpu/samd21/periph/pwm.c [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+#include <string.h>
+
+#include "cpu.h"
+#include "periph/pwm.h"
+#include "periph_conf.h"
+
+/* ignore file in case no PWM devices are defined */
+#if PWM_NUMOF
+
+int round_to_prescaler(unsigned int frequency, unsigned int resolution, int *presc);
```
Comment (thomaseichinger) [inline cpu/samd21/periph/pwm.c]: Could be `static`, right?

File: cpu/samd21/periph/pwm.c [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+            channels = PWM_1_CHANNELS;
+            pwm_clk_id = PWM_1_CLK_ID;
+            pwm_clk_ref = PWM_1_REF_F;
+            break;
+#endif
+    }
+
+    /* setup pins */
+    for (int i = 0; i < channels; i++) {
+        pwm_port->DIRSET.reg = (1 << pwm_pins[i]);  /* tx's direction is output */
```
Comment (thomaseichinger) [inline cpu/samd21/periph/pwm.c]: `tx`?

File: cpu/samd21/periph/pwm.c [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+
+    /* set PWM mode */
+    switch (mode) {
+        case PWM_LEFT:
+            break;
+        case PWM_RIGHT:
+            break;
+        case PWM_CENTER:
+            break;
+        //default:
```
Comment (thomaseichinger) [inline cpu/samd21/periph/pwm.c]: Remove when not needed.

**Comment by PeterKietzmann** [comment]
addressed latest comments

File: boards/samr21-xpro/include/periph_conf.h [hunk 4]
```diff
@@ -42,7 +43,37 @@ extern "C" {
...
 
+/* PWM 1 device configuration */
+#define PWM_1_DEV           TCC1
+#define PWM_1_CHANNELS      2
+#define PWM_1_REF_F         (8000000UL)
+#define PWM_1_CLK_ID        TCC1_GCLK_ID
+#define PWM_1_PM            PM_APBCMASK_TCC1
+/* PWM 1 pin configuration */
+#define PWM_1_PORT          (PORT->Group[0])
+#define PWM_1_PIN_CH0       (6)
```
Comment (PeterKietzmann) [inline boards/samr21-xpro/include/periph_conf.h]: We should be aware that this pin is used for `ADC_0` in PR #2063

**Comment by thomaseichinger** [comment]
@PeterKietzmann needs rebase.

**Comment by PeterKietzmann** [comment]
I rebased and squashed. We should not forget that the `pwm_mode_t mode` is still not handled

**Comment by PeterKietzmann** [comment]
@thomaseichinger I tried to implement the left and right aligned pwm modes for the samr21-xpro. According to the reference manual (section 28.6.2.5 page 617) I wanted to use the up-counting mode for the left-aligned pwm and the down-counting mode for the right-aligned pwm. The corresponding register-content (`PWM_0_DEV->CTRLBCLR.reg` or `PWM_0_DEV->CTRLBSET.reg`) seems to be correct in both cases. However, running the `REMOVE_ME_pwm_test` with a connected logic analyzer shows nearly no difference between the modes. Any ideas?

**Comment by PeterKietzmann** [comment]
In addition: Do you know a way to measure the regarding clock signal?

**Comment by haukepetersen** [comment]
@PeterKietzmann: I just opened a new PR with a slightly adapted version of your PWM driver: #3127. Should we close this then?

**Comment by PeterKietzmann** [comment]
Yes of course! Will test the new one soon

**Pull request closed without merging**